components/suggest/src/store.rs

```diff
@@ -0,0 +1,421 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    ops::Deref,
+    path::{Path, PathBuf},
+};
+
+use once_cell::sync::OnceCell;
+use remote_settings::{self, Attachment, GetItemsOptions, RemoteSettingsConfig, SortOrder};
+use serde_derive::*;
+
+use crate::{
+    db::{ConnectionType, SuggestDb, LAST_FETCH_META_KEY},
+    RemoteRecordId, RemoteSuggestion, Result, SuggestApiResult, Suggestion, SuggestionQuery,
+};
+
+const RS_COLLECTION: &str = "quicksuggest";
+
+/// The store is the entry point to the Suggest component. It incrementally
+/// fetches suggestions from the Remote Settings service, stores them in a local
+/// database, and returns them in response to user queries.
+///
+/// Your application should create a single store, and manage it as a singleton.
+/// The store is thread-safe, and supports concurrent fetches and ingests. We
+/// expect that your application will call `query` to show suggestions as the
+/// user types into the address bar, and periodically call `ingest` in the
+/// background to update the database with new suggestions from Remote Settings.
+///
+/// The store keeps track of the state needed to support incremental ingestion,
+/// but doesn't schedule the ingestion work itself, or decide how many
+/// suggestions to ingest at once. This is for two reasons:
+///
+/// 1. The primitives for scheduling background work vary between platforms. You
+///    might use an idle timer on Desktop, `WorkManager` on Android, or
+///    `BGTaskScheduler` on iOS.
+/// 2. Ingestion limits can change, depending on the platform and the needs of
+///    the application. A mobile device on a metered connection might want to
+///    request a small subset of the Suggest data and fetch the rest later,
+///    while a desktop on a fast link might request the entire dataset on first
+///    launch.
+pub struct SuggestStore {
+    path: PathBuf,
+    dbs: OnceCell<SuggestStoreDbs>,
+    settings_client: remote_settings::Client,
+}
+
+/// Limits for an ingestion from Remote Settings.
+pub struct IngestLimits {
+    /// The maximum number of records to request from Remote Settings.
+    /// Each record has about 200 suggestions.
+    pub records: Option<u64>,
```

Comment:
💭 I added this constraint mostly for smoke testing against a real Remote Settings, so that I could kick the tires on the schema without ingesting the entire 7 MB dataset every time...

But, I think "approximate number of suggestions" might be a more useful constraint than number of records, if it's valuable at all. "Number of records" (and chunks) is an implementation detail; if the consumer requests about 500 suggestions, we can do the calculation in the component that it's 3 records, without leaking that detail to consumers.

Reply:
+1 to hiding the details of what a `record` is - I like the idea of constraints in that it helps for Mobile vs Desktop network connection, but I wonder how the consumer will make the decision of the constraint.. is the plan to run some smoke tests with real devices on throttled networks to see how long the network takes? Or do we plan to have dynamic configuration of limits based on a user's network 👀 

Regardless I feel like this is a followup but I'm interested in the methodology to pick a limit as I'm sure we can learn from that in other places (Nimbus, other consumers of Remote Settings, Sync?)

Reply:
> Or do we plan to have dynamic configuration of limits based on a user's network 👀

Yep, I was thinking something closer to this. I'd love to explore generalizing it to our other components, too, but definitely in the far future!

My half-baked idea is, a consumer could use platform APIs to query whether the device is on a metered connection (and/or has Data Saver or Low Data Mode on), or to check its battery level, and then pick a set of 👋🏼 arbitrary but sensible 👋🏼 limits based on that.

Taking that even further, we could make the limits adaptive: see how long it takes to download some data, and figure out what to prioritize and how much to download based on that.

I'm not sure it's super helpful for where we are right now, since Suggest is a ~7 MB, infrequently changing dataset (1-2x/month; when it does change, it's only a few chunks, not the entire dataset) that's chunked arbitrarily now.

But I could see us wanting to do this if we ever want to support novel suggestion types, like location-based suggestions, in the future—where mirroring the entire dataset might not work well.

How does that line up with what you're thinking? 😊

Reply:
Didn't realize I didn't get back to this 😅
> My half-baked idea is, a consumer could use platform APIs to query whether the device is on a metered connection (and/or has Data Saver or Low Data Mode on), or to check its battery level, and then pick a set of 👋🏼 arbitrary but sensible 👋🏼 limits based on that.

 I like it! And seems like it would serve our use case here of making sure the user gets a good experience even if they are on a low data network connection - once we ship and see how well our configuration story is in practice I'd love for us to visit if there's value in the other components, but like you said, that's in the far future!

Reply:
> My half-baked idea is, a consumer could use platform APIs to query whether the device is on a metered connection (and/or has Data Saver or Low Data Mode on), or to check its battery level, and then pick a set of 👋🏼 arbitrary but sensible 👋🏼 limits based on that.

I guess one vague concern about this is that "arbitrary but sensible" is really a contradiction. Picking something sensible probably means the app needs to know more about the implementation than really makes sense. I'm wondering if it's better for the app to communicate what it does know for sure (ie, it's in a kind of data-saver mode) which the component uses to choose appropriate (but not arbitrary ;) limits.